kotlinx-coroutines-core/jvm/src/debug/internal/DebugCoroutineInfoImpl.kt

```diff
@@ -40,19 +40,101 @@ internal class DebugCoroutineInfoImpl(
      * Can be CREATED, RUNNING, SUSPENDED.
      */
     public val state: String get() = _state
+
+    @Volatile
     private var _state: String = CREATED
 
+    /*
+     * How many consecutive unmatched 'updateState(RESUMED)' this object has received.
+     * It can be `> 1` in two cases:
+     *
+     * * The coroutine is finishing and its state is being unrolled in BaseContinuationImpl, see comment to DebugProbesImpl#callerInfoCache
+     *   Such resumes are not expected to be matched and are ignored.
+     * * We encountered suspend-resume race explained above, and we do wait for a match.
+     */
+    private var unmatchedResume = 0
+
+    /**
+     * Here we orchestrate overlapping state updates that are coming asynchronously.
+     * In a nutshell, `probeCoroutineSuspended` can arrive **later** than its matching `probeCoroutineResumed`,
+     * e.g. for the following code:
+     * ```
+     * suspend fun foo() = yield()
+     * ```
+     *
+     * we have this sequence:
+     * ```
+     * fun foo(...) {
+     *     uCont.intercepted().dispatchUsingDispatcher() // 1
+     *     // Notify the debugger the coroutine is suspended
+     *     probeCoroutineSuspended() // 2
+     *     return COROUTINE_SUSPENDED // Unroll the stack
```

Comment:
Is it correct that (though astronomically unlikely?) `probeCoroutineSuspended` can even come out of order, and the fix doesn't prevent that? Let's say that, while the first `probeCoroutineSuspended` was awaiting execution, the coroutine was suspended again, and the new `probeCoroutineSuspended` managed to happen before the first one did. So, could we discard the wrong `probeCoroutineSuspend` here?